### PR TITLE
GM_config as constructor

### PR DESCRIPTION
**Pull request by sizzlemctwizzle**
GM_config is now the constructor. Backwards compatibility is seamless. Passes the unit test script without modification.

Now this works:
``` js
let config = new GM_config(
{
  'id': 'MyConfig', // The id used for this instance of GM_config
  'fields': // Fields object
  {
    'Name': // This is the id of the field
    {
      'label': 'Name', // Appears next to field
      'type': 'text', // Makes this setting a text field
      'default': 'Sizzle McTwizzle' // Default value if user doesn't change it
    }
  }
});
```

### CODE REVIEW

File: gm_config.js [hunk 1]
```diff
@@ -44,509 +44,484 @@ GM_config is distributed under the terms of the GNU Lesser General Public Licens
...
-        "#GM_config .section_header { background: #414141; border: 1px solid #000; color: #FFF;",
-        " font-size: 13pt; margin: 0; }",
-        "#GM_config .section_desc { background: #EFEFEF; border: 1px solid #CCC; color: #575757;" +
-          " font-size: 9pt; margin: 0 0 6px; }"
-        ].join('\n') + '\n',
-      basicPrefix: "GM_config",
-      stylish: ""
-    };
-  }
+let GM_config = (function () {
```
Comment (Martii) [inline gm_config.js:47]: I think there's going to be a breaking change here with some implementation... [Where I do this](https://github.com/Martii/UserScripts/blob/be98c56f53a184f1e88245cbc2c819990621925f/src/oujs/OUJS-1/OUJS-1.user.js#L45-L47) for garbage cleanup will have to go for sure because of...

File: gm_config.js [hunk 2]
```diff
@@ -44,509 +44,484 @@ GM_config is distributed under the terms of the GNU Lesser General Public Licens
...
+  config.setValue = construct.prototype.setValue;
+  config.getValue = construct.prototype.getValue;
+  config.stringify = construct.prototype.stringify;
+  config.parser = construct.prototype.parser;
+  config.log = construct.prototype.log;
+  config.remove = construct.prototype.remove;
+
+  return config;
+}());
+let GM_configStruct = GM_config;
```
Comment (Martii) [inline gm_config.js:524]: ... this as well. I don't mind changing it but definitely will be breaking one implementation.

**Review by Martii — COMMENTED** [review]
Don't have a huge amount of time atm to review in depth but do see this as a potential breaking change at the commented. I'll try to get around to retesting it on my end hopefully this weekend but if not... mid next week.

**Comment by Martii** [comment]
Just a misc question to... Milestone 4.0? Versioning?

File: gm_config.js (hunk 2 above)
Comment (sizzlemctwizzle) [reply gm_config.js:524]: Line 524 should preserve the `GM_configStruct` usage, including where `GM_config` is overwritten, like how you set it to `undefined`. Am I missing something that makes this not possible?

**Comment by sizzlemctwizzle** [comment]
The previous milestone was called 3.0 so I just went with 4.0. No real deep thought there.

File: gm_config.js (hunk 2 above)
Comment (sizzlemctwizzle) [reply gm_config.js:524]: Confirmed. This works as expected:
```js
GM_config = undefined;

var gmc = new GM_configStruct();

gmc.init(
{
  'id': 'MyConfig', // The id used for this instance of GM_config
  'fields': // Fields object
  {
    'Name': // This is the id of the field
    {
      'label': 'Name', // Appears next to field
      'type': 'text', // Makes this setting a text field
      'default': 'Sizzle McTwizzle' // Default value if user doesn't change it
    }
  }
});

gmc.open();
```

**Comment by sizzlemctwizzle** [comment]
There is still an issue with this branch. `getValue` won't work with GM4+ because it currently needs to be synchronous. I'm just going to copy the pollyfill for this one and modify `read` to be async.

File: gm_config.js [hunk 3]
```diff
@@ -148,20 +148,21 @@ let GM_config = (function () {
 
     // Create the fields
     if (settings.fields) {
-      var stored = config.read(), // read the stored settings
-          fields = settings.fields,
+      config.read(null, (stored) => { // read the stored settings
```
Comment (Martii) [inline gm_config.js:151]: Breaking change and partial migration from ES 5.x to ES 6 here and a few other places that use the [arrow function expression](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Arrow_functions).

File: gm_config.js [hunk 4]
```diff
@@ -479,9 +479,17 @@ let GM_config = (function () {
   
   construct.prototype.name = 'GM_config';
   construct.prototype.constructor = construct;
+  let isGM4 = typeof GM === 'object';
```
Comment (Martii) [inline gm_config.js]: ES 5.0 migration to at minimum ES 5.1... or ES6 for non Moz browsers.

File: gm_config.js [hunk 5]
```diff
@@ -481,11 +481,11 @@ let GM_config = (function () {
   construct.prototype.constructor = construct;
   let isGM4 = typeof GM === 'object';
   if (isGM4) {
-    let GM_getValue = (name, def) => await GM.getValue(name, def);
-    let GM_setValue = (name, value) => await GM.setValue(name, value);
-    let GM_log = (text) => await GM.log(text);
+    let GM_getValue = async (name, def) => await GM.getValue(name, def);
+    let GM_setValue = async (name, value) => await GM.setValue(name, value);
```
Comment (Martii) [inline gm_config.js]: This seems to have been done earlier... but abandoning synchronous ability. Note TM doesn't "officially support" GM4 API and you'll be breaking SM GM Port because it's v3 only. Once Manifest v3 come full swing I'm not sure what Mozilla will do with GM itself as far as compatibility... still up in the air from everything I've read.

File: gm_config.js (hunk 2 above)
Comment (Martii) [reply gm_config.js:554]: OUJS-1 breaks with the current gm_config.js... I'll have to see what's going on. See https://openuserjs.org/user/preferences under the OUJS-1 header.

Comment (Martii) [reply gm_config.js:554]: Console error:

``` console
14:55:04.194 Script error in [Greasemonkey script https://openuserjs.org/users/Marti/OUJS-1; version 2.0.0a.10]:
TypeError: field is undefined OUJS-1:1168:19
```

Comment (Martii) [reply gm_config.js:554]: Chromium/TM console:

``` console
caught (in promise) TypeError: Cannot read properties of undefined (reading 'value')
    at construct.get (userscript.html?name=OUJS-1.user.js&id=51039475-70bc-444b-9f5d-e980a65c8acf:387:50)
    at userscript.html?name=OUJS-1.user.js&id=51039475-70bc-444b-9f5d-e980a65c8acf:987:11
    at Proxy.<anonymous> (userscript.html?name=OUJS-1.user.js&id=51039475-70bc-444b-9f5d-e980a65c8acf:1103:3)
    at St (<anonymous>:9:89)
    at userscript.html?name=OUJS-1.user.js&id=51039475-70bc-444b-9f5d-e980a65c8acf:1:87
    at window.__f__lhe8jycy.nv (userscript.html?name=OUJS-1.user.js&id=51039475-70bc-444b-9f5d-e980a65c8acf:1:359)
    at St (<anonymous>:9:89)
    at n (<anonymous>:71:248)
    at <anonymous>:74:85
    at p (<anonymoss>:68:59)
```

Ref: 
* https://github.com/sizzlemctwizzle/GM_config/blob/f574c2cb8564301725ead3d3c3cb1bc83ed0d647/gm_config.js#L386

Misc note:

* `GM_log` throws an error now in Chromium/TM without `@grant` for it. :\

Comment (sizzlemctwizzle) [reply gm_config.js:554]: Yeah something was wrong with my dev environment. This branch still needs a _lot_ of work.

**Comment by Martii** [comment]
No problem... I look forward to your next changes. Please note I'm not against an ES migration but it needed to be said for those that are still using more ancient OS's that are stuck on older browsers.

**Comment by Martii** [comment]
@sizzlemctwizzle 

One other quickie note too.. when you assigned me to review this... GH did **not** notify me... just the usual subscription method of notifying there was a new PR and such. So I'm guessing you need to @ mention me for a confirmation.

**Comment by sizzlemctwizzle** [comment]
Reworked the branch and got everything working, except the unit test script won't work on the translation example because of the breaking change with GM_config.getValue. So changed my mind and reopening this PR.

**Comment by sizzlemctwizzle** [comment]
Okay, @Martii, this is finally ready to be reviewed. I have an [updated version](https://gist.github.com/sizzlemctwizzle/a4bedc76ffcf01a3db500099833df02a) of the unit test script for testing purposes. I only made a minimal change to make it work the new async version of `GM_config.getValue`.

**Comment by Martii** [comment]
Still getting with [this current PR gmconfig.js](https://github.com/sizzlemctwizzle/GM_config/blob/376803bc507768c0ed2204d560504ec05072348c/gm_config.js):

``` console
20:44:30.729 Script error in [Greasemonkey script https://openuserjs.org/users/Marti/OUJS-1; version 2.0.0a.10]:
TypeError: field is undefined OUJS-1:1176:19
    <anonymous> user-script:https://openuserjs.org/users/Marti/OUJS-1:1176
```

and 

``` console
userscript.html?name=OUJS-1.user.js&id=51039475-70bc-444b-9f5d-e980a65c8acf:397 Uncaught (in promise) TypeError: Cannot read properties of undefined (reading 'value')
    at construct.get (userscript.html?name=OUJS-1.user.js&id=51039475-70bc-444b-9f5d-e980a65c8acf:397:50)
    at userscript.html?name=OUJS-1.user.js&id=51039475-70bc-444b-9f5d-e980a65c8acf:995:11
    at Proxy.<anonymous> (userscript.html?name=OUJS-1.user.js&id=51039475-70bc-444b-9f5d-e980a65c8acf:1111:3)
    at St (<anonymous>:9:89)
    at userscript.html?name=OUJS-1.user.js&id=51039475-70bc-444b-9f5d-e980a65c8acf:1:88
    at window.__f__lhhahhez.kyn (userscript.html?name=OUJS-1.user.js&id=51039475-70bc-444b-9f5d-e980a65c8acf:1:360)
    at St (<anonymous>:9:89)
    at n (<anonymous>:71:248)
    at <anonymous>:74:85
    at p (<anonymous>:68:59)
```

... respectively... however the actual options now show up in OUJS-1 but they are failing out when being applied. I am only selecting "More emphasis on current username" then go to all discussions and hunt down my user name of whatever user I'm logged in with... that should be black background instead of the normal blue.

Same exception location as before of:

https://github.com/sizzlemctwizzle/GM_config/blob/376803bc507768c0ed2204d560504ec05072348c/gm_config.js#L396-L397


This is the only change I've actually made locally to [OUJS-1](https://openuserjs.org/scripts/Marti/OUJS-1) for GM vs GM_ possible compatibility and weird TM behavior:

``` diff
--- /scripts/Marti/OUJS-1/source@2.0.0a.10+337d73b
+++ /scripts/Marti/OUJS-1/source
@@ -28,6 +28,9 @@
 
 // @grant  GM_getValue
 // @grant  GM_setValue
+// @grant  GM.getValue
+// @grant  GM.setValue
+// @grant  GM_log
 
 // @noframes
```
 Also the Unit Test gist seems to only test the GM_* i.e. probably should have at least two test files to make sure both logic paths are fully tested with GM.* and GM_* with the `@grant`s?

**Comment by Martii** [comment]
This enables it to apply the emphasize username:

``` diff
--- /scripts/Marti/OUJS-1/source@2.0.0a.10+337d73b
+++ /scripts/Marti/OUJS-1/source
@@ -28,6 +28,8 @@
 
 // @grant  GM_getValue
 // @grant  GM_setValue
+// @grant  GM.getValue
+// @grant  GM.setValue
 
 // @noframes
 
@@ -119,6 +121,8 @@
     gmc.open();
   }
 
+  alert('delay so open can finish');
+  
   if (gmc.get('showScriptListLineNumbers')) {
     var count = null;
 
@@ -201,6 +205,8 @@
   }
 
   if (gmc.get('emphasizeCurrentUsername')) {
+    
+    alert('yes it is set');
 
     var nodeStyle = document.createElement('style');
     nodeStyle.setAttribute("type", "text/css");
```

i.e. gives it more time for `.open` to finish I believe. Stopping here until you give feedback or perhaps a change.

**Comment by sizzlemctwizzle** [comment]
@Martii 
```js
   if (gmc.get('showScriptListLineNumbers')) {
```
Fields will not be loaded yet since `init` is now async. This should be done in the `onInit` callback.

> the Unit Test gist seems to only test the GM_* i.e. probably should have at least two test files to make sure both logic paths are fully tested with GM.* and GM_* with the @grants?

Shouldn't be any difference. There is a pollyfill for GM_* (same for localStorage) that makes it work like GM.*

**Comment by Martii** [comment]
> Fields will not be loaded yet since `init` is now async.

Good to know... although bear with me on this question please... Where exactly is the change from synchronous to asynchronous? I see the read/write one but nothing else from skims. Would be helpful to denote this as well in [the unfinished section](https://github.com/sizzlemctwizzle/GM_config/wiki/Internal-Documentation-Index) of the wiki too. i.e. what is synchronous vs asynchronous.

> Backwards compatibility is seamless. Passes the unit test script without modification.

And `gmc.open`. The GM_config Unit Test is also affected albeit not throwing the error **yet**.  If `open` is called before `init` is finished then it too will get the errors.

> Shouldn't be any difference.

Doesn't seem to be in SM... although do have a compatibility issue with the extension and latest SM. Ughh... that'll have to be next week for diagnosis for sure.

> There is a pollyfill for GM_*

Thanks. My skim *(limited time)* gathered that but needed confirmation in plain english.

> This should be done in the `onInit` callback.

OUJS-1 updated to reflect the upcoming change. It is [current  HEAD GMC](https://github.com/sizzlemctwizzle/GM_config/blob/2207c5c1322ebb56e401f03c2e581719f909762a/gm_config.js) and [new PR GMC](https://github.com/sizzlemctwizzle/GM_config/blob/02b49098cfecaa1469ae4e3a9280582a64fdbbc6/gm_config.js) compatible.

As far as my time for examining more details... it's run out until early next week. :\ Have a trip to attend to in the morning.

We'll also have to see if @MysteryBlokHed is available to update the type definitions *after* it's merged.

**Comment by sizzlemctwizzle** [comment]
> Where exactly is the change from synchronous to asynchronous?
`getValue`, `setValue`, `log`, `read`, `write`, `save`, and `init` are all asynchronous now (`open` was already asynchronous). But I consider `getValue`, `setValue`, `log`, `read`, and `write` to be internal so I don't count them as a breaking change. `save` and `init` are the official API functions that are effected by a breaking change. I will update the wiki so that examples put effected code in either the `save` or `init` event callbacks.

**Comment by Martii** [comment]
Hmm... I seem to be getting `load` called twice both in SM and Fx. Would you double check that please.

Usage:
* https://github.com/sizzlemctwizzle/GM_config/blob/02b49098cfecaa1469ae4e3a9280582a64fdbbc6/gm_config.js
* https://gist.github.com/sizzlemctwizzle/a4bedc76ffcf01a3db500099833df02a/9d24f8dcbd4395ad92f713a72c29d08a9b96995e

Exception:
* First run *(this may be trapped when needed but I think the dual `load` trigger might be an issue)*

``` console
10:33:17.898 TypeError: frame.contentDocument is null sandbox eval code:750:24
    open user-script:sizzlemctwizzle/The GM_config Unit Test:750
```
* Second run

Okay

Ref:
* #98

**Comment by Martii** [comment]
Also get this in TM:

``` console
erscript.html?name=OUJS-1.user.js&id=51039475-70bc-444b-9f5d-e980a65c8acf:397 Uncaught (in promise) TypeError: Cannot read properties of undefined (reading 'value')
    at construct.get (userscript.html?name=OUJS-1.user.js&id=51039475-70bc-444b-9f5d-e980a65c8acf:397:50)
    at userscript.html?name=OUJS-1.user.js&id=51039475-70bc-444b-9f5d-e980a65c8acf:995:11
    at Proxy.<anonymous> (userscript.html?name=OUJS-1.user.js&id=51039475-70bc-444b-9f5d-e980a65c8acf:1111:3)
    at St (<anonymous>:9:89)
    at userscript.html?name=OUJS-1.user.js&id=51039475-70bc-444b-9f5d-e980a65c8acf:1:87
    at window.__f__lhs5ca3b.ty (userscript.html?name=OUJS-1.user.js&id=51039475-70bc-444b-9f5d-e980a65c8acf:1:359)
    at St (<anonymous>:9:89)
    at n (<anonymous>:71:248)
    at <anonymous>:74:85
    at p (<anonymous>:68:59)
```

... which points to *(2nd line in message):

https://github.com/sizzlemctwizzle/GM_config/blob/02b49098cfecaa1469ae4e3a9280582a64fdbbc6/gm_config.js#L396-L397

Neither of these seem fatal but is questionable.

**Comment by sizzlemctwizzle** [comment]
Could reproduce the this error in Fx with GM:
```console
12:33:17.898 TypeError: frame.contentDocument is null sandbox eval code:750:24
    open user-script:sizzlemctwizzle/The GM_config Unit Test:750
```
And I just fixed it with my latest commit.

Still haven't figured what's going on with [OUJS-1](https://gist.github.com/sizzlemctwizzle/7d6970d20e3da463feff9b6e83392bf8) in Chrome with TM. Now I get this puzzling error:
```console
Uncaught (in promise) ReferenceError: GM_log is not defined
    at userscript.html?name=OUJS-1.user.js&id=06ee5f0a-972c-4c16-9dc2-b5d058d8d3d0:521:22
    at Proxy.<anonymous> (userscript.html?name=OUJS-1.user.js&id=06ee5f0a-972c-4c16-9dc2-b5d058d8d3d0:560:2)
    at St (<anonymous>:9:89)
    at userscript.html?name=OUJS-1.user.js&id=06ee5f0a-972c-4c16-9dc2-b5d058d8d3d0:1:87
    at window.__f__lhuxscx9.bt (userscript.html?name=OUJS-1.user.js&id=06ee5f0a-972c-4c16-9dc2-b5d058d8d3d0:1:350)
    at St (<anonymous>:9:89)
    at n (<anonymous>:71:248)
    at <anonymous>:74:85
    at p (<anonymous>:68:59)
```

**Comment by Martii** [comment]
Bleh.. debugging TM is just not my favorite cup of tea... not to mention Chromiums issue with copy/paste in their messages.

Anyhow... long story short the error has cleared itself after a purge of the Chromium profile and reinstalled fresh.

Now that I'm sufficiently brain dead from a corrupt profile... I need a break... I don't see any other errors runtime so far.

**Comment by sizzlemctwizzle** [comment]
> I need a break... I don't see any other errors runtime so far.

I don't either, so I'll merge this now and we can just deal with any issues that crop up in `master`.

**Comment by sizzlemctwizzle** [comment]
I incorrectly thought something went wrong after the merge so I backed out the changes. Then when I came to my senses, I re-reverted. Anyway, I'm done with this PR. Any bugs we might find should have their own issue from now on. *mic drop*
